org.eclipse.acute/src/org/eclipse/acute/dotnetnew/DotnetNewWizard.java

```diff
@@ -101,10 +131,83 @@ public boolean performFinish() {
 				}
 			}
 		} catch (CoreException e) {
-			MessageDialog.openError(getShell(), "Unable to load project description", "");
+			MessageDialog.openError(getShell(), "Error", "Unable to load project description.");
 			return false;
 		}
 		return true;
 	}
 
+	private Set<IWorkingSet> getWorkingSets(Object o) {
+		IWorkingSet[] allWorkingSets = PlatformUI.getWorkbench().getWorkingSetManager().getAllWorkingSets();
+		Set<IWorkingSet> fileWorkingSets = new HashSet<>();
+
+		for (IWorkingSet iWorkingSet : allWorkingSets) {
+			IAdaptable[] elements = iWorkingSet.getElements();
+
+			for (IAdaptable element : elements) {
+				if (element.equals(o)) {
+					fileWorkingSets.add(iWorkingSet);
+					break;
+				} else if (element instanceof IContainer) {
+					IContainer container = (IContainer) element;
+					boolean objectFound = false;
+
+					for (IResource resource : recursiveResources(container)) {
```

Comment:
Recursing here can be very very expensive. You can rely on the assumption that working sets most likely contain IProject -or things that adapt to it.
So Assuming you pass an IResource, then finding whether a working set is related can be simply `Arrays.asList(elements).contains(myResource.getProject())`, which is a `O(sizeof(workingSet))` operation, in a O(number_of_workingSets) operation)`, where we know that `sizeof(workingSet) <= number_of_projects`, so the method would be `O(number_of_workingSets*number_of_projects)` which is way cheaper than the current one which is `O(number_of_workingSets*number_of_IResources)`.